### PR TITLE
Address security issues

### PR DESCRIPTION
**Pull request by ryanmelt**
Addresses:
GHSL-2024-127
GHSL-2024-128
GHSL-2024-129

### CODE REVIEW

File: openc3-cosmos-cmd-tlm-api/app/controllers/application_controller.rb [hunk 1]
```diff
@@ -57,4 +57,34 @@ def authorization(permission, target_name: nil)
...
+        # This method is taken directly from the Rails source:
+        #   https://api.rubyonrails.org/v5.2/classes/ActiveStorage/Filename.html#method-i-sanitized
+        if allow_forward_slash
+          # Sometimes we have forward slashes so optionally allow those
+          value = arg.encode(Encoding::UTF_8, invalid: :replace, undef: :replace, replace: "�").strip.tr("\u{202E}%$|:;\t\r\n\\", "-")
+        else
+          value = arg.encode(Encoding::UTF_8, invalid: :replace, undef: :replace, replace: "�").strip.tr("\u{202E}%$|:;/\t\r\n\\", "-")
+        end
+        if value != arg
+          render(json: { status: 'error', message: "Invalid parameter #{param_list[index]}" }, status: 400)
```
Comment (jmthomas) [inline openc3-cosmos-cmd-tlm-api/app/controllers/application_controller.rb:83]: Should this still be `param_list[index]` or should it be `result[index]`?

File: openc3-cosmos-cmd-tlm-api/app/controllers/storage_controller.rb [hunk 2]
```diff
@@ -25,6 +25,7 @@
 
 class StorageController < ApplicationController
   def buckets
+    return unless authorization('system')
```
Comment (jmthomas) [inline openc3-cosmos-cmd-tlm-api/app/controllers/storage_controller.rb:28]: This is a good check but doesn't everyone have `system`?

File: openc3-cosmos-init/plugins/openc3-tool-base/public/js/auth.js [hunk 3]
```diff
@@ -35,9 +35,14 @@ class Auth {
     delete localStorage.openc3Token
   }
   login(redirect) {
+    let url = new URL(redirect)
+    let result = url.pathname
+    if (url.search) {
+      result = result + url.search
+    }
```
Comment (jmthomas) [inline openc3-cosmos-init/plugins/openc3-tool-base/public/js/auth.js:42]: What are you accomplishing here? Comment?

File: openc3-cosmos-init/plugins/packages/openc3-tool-common/src/tools/base/components/Login.vue [hunk 4]
```diff
@@ -145,7 +145,12 @@
       const redirect = new URLSearchParams(window.location.search).get(
         'redirect',
       )
-      window.location = decodeURI(redirect || '/')
+      if (redirect[0] === '/' && redirect[1] !== '/') {
+        // Valid relative redirect URL
+        window.location = decodeURI(redirect)
```
[the remark this answers is not in the document]
Comment (jmthomas) [reply openc3-cosmos-init/plugins/packages/openc3-tool-common/src/tools/base/components/Login.vue:150]: Could we whitelist the redirects like this comment says with something like `['/tool/cmdtlmserver','tool/scriptrunner', ...]`

**Review by jmthomas — COMMENTED** [review]
Looks good with a lot of changes. Not sure if there's something we can do about the remaining issues here. Does this cover both CVEs?

File: openc3-cosmos-cmd-tlm-api/app/controllers/application_controller.rb (hunk 1 above)
Comment (ryanmelt) [reply openc3-cosmos-cmd-tlm-api/app/controllers/application_controller.rb]: I updated to match other messages.

File: openc3-cosmos-cmd-tlm-api/app/controllers/storage_controller.rb (hunk 2 above)
Comment (ryanmelt) [reply openc3-cosmos-cmd-tlm-api/app/controllers/storage_controller.rb:28]: Yes, but at least not it requires you to be authenicated.  Before even not logged in could get it.

File: openc3-cosmos-init/plugins/openc3-tool-base/public/js/auth.js (hunk 3 above)
Comment (ryanmelt) [reply openc3-cosmos-init/plugins/openc3-tool-base/public/js/auth.js:42]: This converts to a relative url.

**Pull request merged**